packages/react-sdk/src/state/communication/discovery/matrixRtcSessionManagerImpl.ts

```diff
@@ -0,0 +1,386 @@
+  private readonly sessionSubject = new Subject<SessionState>();
+  private sessions: RTCSessionEventContent[] = [];
+  private joinState: { whiteboardId: string; sessionId: string } | undefined;
+
+  constructor(
+    private readonly widgetApiPromise: Promise<WidgetApi> | WidgetApi,
+    private readonly sessionTimeout = DEFAULT_RTC_EXPIRE_DURATION,
+  ) {}
+
+  getSessionId(): string | undefined {
+    return this.joinState?.sessionId;
+  }
+
+  getSessions(): Session[] {
+    return this.sessions.map(({ session_id, user_id }) => ({
+      sessionId: session_id,
+      userId: user_id,
+    }));
+  }
+
+  observeSessionJoined(): Observable<Session> {
+    return this.sessionJoinedSubject;
+  }
+
+  observeSessionLeft(): Observable<Session> {
+    return this.sessionLeftSubject;
+  }
+
+  observeSession(): Observable<SessionState> {
+    return this.sessionSubject;
+  }
+
+  async join(whiteboardId: string): Promise<{ sessionId: string }> {
+    if (this.joinState) {
+      this.logger.debug('Already joined a whiteboard, must leave first.');
+      await this.leave();
+    }
+
+    const widgetApi = await this.widgetApiPromise;
+    const { userId, deviceId } = widgetApi.widgetParameters;
+    const sessionId = `_${userId}_${deviceId}`;
+
+    this.joinState = { sessionId, whiteboardId };
+
+    this.logger.debug(
+      `Joining whiteboard ${whiteboardId} as session ${sessionId}`,
+    );
+
+    // Handle session events
+    from(Promise.resolve(this.widgetApiPromise))
+      .pipe(
+        switchMap((widgetApi) =>
+          widgetApi.observeStateEvents(STATE_EVENT_RTC_MEMBER),
+        ),
+        filter(isValidRTCSessionStateEvent),
+        takeUntil(this.destroySubject),
+        takeUntil(this.leaveSubject),
+      )
+      .subscribe((rtcSession) => {
+        this.handleRTCSessionEvent(rtcSession);
+      });
+
+    interval(this.sessionTimeout * 0.75)
+      .pipe(
+        takeUntil(this.destroySubject),
+        takeUntil(this.leaveSubject),
+        switchMap(() => this.refreshOwnSession(whiteboardId, sessionId)),
+      )
+      .subscribe();
+
+    await this.refreshOwnSession(whiteboardId, sessionId);
+
+    return { sessionId };
+  }
+
+  async leave(): Promise<void> {
+    if (!this.joinState) {
+      return;
+    }
+    const { sessionId, whiteboardId } = this.joinState;
+
+    this.joinState = undefined;
+    this.leaveSubject.next();
+
+    this.logger.log(
+      `Leaving whiteboard ${whiteboardId} as session ${sessionId}`,
+    );
+
+    const widgetApi = await this.widgetApiPromise;
+    const { userId } = widgetApi.widgetParameters;
+
+    this.removeSession({ state_key: sessionId, sender: userId });
+
+    await this.removeOwnSession(whiteboardId, sessionId);
+  }
+
+  destroy(): void {
+    this.destroySubject.next();
+    this.sessionJoinedSubject.complete();
+    this.sessionSubject.complete();
+    this.sessionLeftSubject.complete();
+  }
+
+  private handleRTCSessionEvent(
```

Comment:
I think it should be possible to simplify code that handles updates and additions of events, maybe some more tests could be added.